tapdb/sqlite.go

```diff
@@ -149,6 +157,87 @@ func NewSqliteStore(cfg *SqliteConfig) (*SqliteStore, error) {
 	return s, nil
 }
 
+// backupSqliteDatabase creates a backup of the given SQLite database.
+func backupSqliteDatabase(srcDB *sql.DB, dbFullFilePath string) error {
+	if srcDB == nil {
+		return fmt.Errorf("backup source database is nil")
+	}
+
+	// Create a database backup file full path from the given source
+	// database full file path.
+	//
+	// Get the current time and format it as a Unix timestamp in
+	// nanoseconds.
+	timestamp := time.Now().UnixNano()
+
+	// Add the timestamp to the backup name.
+	backupFullFilePath := fmt.Sprintf(
+		"%s.%d.backup", dbFullFilePath, timestamp,
+	)
+
+	log.Infof("Creating backup of database file: %v -> %v",
+		dbFullFilePath, backupFullFilePath)
+
+	// Backup the database
+	vacuumIntoQuery := "VACUUM INTO ?;"
+	stmt, err := srcDB.Prepare(vacuumIntoQuery)
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	_, err = stmt.Exec(backupFullFilePath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// backupAndMigrate is a helper function that creates a database backup before
+// initiating the migration, and then migrates the database to the latest
+// version.
+func (s *SqliteStore) backupAndMigrate(mig *migrate.Migrate,
+	currentDbVersion fn.Option[uint],
+	maxMigrationVersion fn.Option[uint]) error {
+
+	// Determine if a database migration is necessary given the current
+	// database version and the maximum migration version known to the
+	// driver.
+	versionUpgradePending := false
+	currentDbVersion.WhenSome(func(currentV uint) {
+		maxMigrationVersion.WhenSome(func(maxV uint) {
+			versionUpgradePending = currentV < maxV
+		})
+	})
+
+	// If a database migration version upgrade is pending, create a backup
+	// of the database before starting the migration.
+	if versionUpgradePending && !s.cfg.SkipMigrationDbBackup {
```

Comment:
nit: could make this into a 3-way switch statement to distinguish between `versionUpgradePending && !s.cfg.SkipMigrationDbBackup`,  `versionUpgradePending && s.cfg.SkipMigrationDbBackup` and `!versionUpgradePending`, at least for the log message.

Reply:
I've revised this function. More logical now IMO. Let me know what you think.